src/irr/asset/CSmoothNormalGenerator.cpp

```diff
@@ -0,0 +1,254 @@
+#include "CSmoothNormalGenerator.h"
+
+#include <iostream>
+#include <algorithm>
+#include <array>
+
+namespace irr
+{
+namespace asset
+{
+static inline bool operator<(uint32_t lhs, const IMeshManipulator::SSNGVertexData& rhs)
+{
+	return lhs < rhs.hash;
+}
+
+static inline bool operator<(const IMeshManipulator::SSNGVertexData& lhs, uint32_t rhs)
+{
+	return lhs.hash < rhs;
+}
+
+static inline bool compareVertexPosition(const core::vectorSIMDf& a, const core::vectorSIMDf& b, float epsilon)
+{
+	const core::vectorSIMDf difference = core::abs(b - a);
+	return (difference.x <= epsilon && difference.y <= epsilon && difference.z <= epsilon);
+}
+
+static inline core::vector3df_SIMD getAngleWeight(const core::vector3df_SIMD & v1,
+	const core::vector3df_SIMD & v2,
+	const core::vector3df_SIMD & v3)
+{
+		// Calculate this triangle's weight for each of its three vertices
+		// start by calculating the lengths of its sides
+	const float a = v2.getDistanceFromSQAsFloat(v3);
+	const float asqrt = sqrtf(a);
+	const float b = v1.getDistanceFromSQAsFloat(v3);
+	const float bsqrt = sqrtf(b);
+	const float c = v1.getDistanceFromSQAsFloat(v2);
+	const float csqrt = sqrtf(c);
+
+		// use them to find the angle at each vertex
+	return core::vector3df_SIMD(
+		acosf((b + c - a) / (2.f * bsqrt * csqrt)),
+		acosf((-b + c + a) / (2.f * asqrt * csqrt)),
+		acosf((b - c + a) / (2.f * bsqrt * asqrt)));
+}
+
+asset::ICPUMeshBuffer* irr::asset::CSmoothNormalGenerator::calculateNormals(asset::ICPUMeshBuffer* buffer, float epsilon, asset::E_VERTEX_ATTRIBUTE_ID normalAttrID, IMeshManipulator::VxCmpFunction vxcmp)
+{
+	VertexHashMap vertexArray = setupData(buffer, epsilon);
+	processConnectedVertices(buffer, vertexArray, epsilon, normalAttrID, vxcmp);
+
+	return buffer;
+}
+
+CSmoothNormalGenerator::VertexHashMap::VertexHashMap(size_t _vertexCount, uint32_t _hashTableMaxSize, float _cellSize)
+	:hashTableMaxSize(_hashTableMaxSize), 
+	cellSize(_cellSize)
+{
+	_IRR_DEBUG_BREAK_IF((!core::isPoT(_hashTableMaxSize)));
```

Comment:
this needs to be an actual `assert(!core::isPoT(_hashTableMaxSize))`, we want the program to crash (with a message) 

`_IRR_DEBUG_BREAK_IF` only breaks the debugger if one is used, otherwise execution continues.